### PR TITLE
Automated Data Collection - will be merged once complete for *all* assets

### PR DESCRIPTION
**Pull request by DZGoldman**
Crawls web for BTC and ETH data and writes to yml files.

run it with 
```
python _data/scripts/main.py
```
from the root.

Figured we'd start with these two coins to see if we like this basic structure (working on some other coins with varying degrees of progress).

Note that as of posting this, https://coin.dance/ is undergoing maintenance, and hence those two BTC tasks fail.

### CODE REVIEW

File: _data/scripts/main.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+# Start driver with specified options
+driver = webdriver.Chrome(chrome_options=options) 
+
+bitcoin = Bitcoin(driver)
+bitcoin.main()
+
+eth = Ethereum(driver)
+eth.main()
+
```
Comment (DZGoldman) [inline _data/scripts/main.py]: This'll be made to be more dynamic as we add more coins, including command line arguments for only updating data for certain coins, or whatever is convenient.

**Comment by DZGoldman** [comment]
(Pushed up the XRP script as well)

**Comment by DZGoldman** [comment]
Ty, will update

**Comment by DZGoldman** [comment]
...so actually, the API doesn't offer a way to get the rich list, and the node count always caps at 200 :/

**Comment by MarkusTeufelberger** [comment]
https://developers.ripple.com/data-api-v2-tool.html#get-topology
I get a `node_count` of 800 there...?

The rich list is a separate tool that's not done by Ripple but by @WietseWind

**Comment by WietseWind** [comment]
**XRPL node count**:
https://data.ripple.com/v2/network/topology - `//node_count`
(FYI: [some history](https://nubg3ex.dlvr.cloud/pasted_1.png) of this endpoint+metric from my monitoring)

**XRPL validator count**:
https://data.ripple.com/v2/network/validator_reports - `//count`
(FYI: [some history](https://udoc6hh.dlvr.cloud/pasted_1.png) of this endpoint+metric from my monitoring)

**Richlist**:
https://ledger.exposed/api/richlist
You can find the percentages in the `//pct` node, where `//pct/pct10` is the top 10%, and `//pct/pct0p1` is 0.1% - values in XRP.

The total supply of available XRP is either `//totalCoins` (all coins, incl. escrows) or the sum of all `//has/?/balanceSum` entries (excl. escrows)

**Comment by DZGoldman** [comment]
ty!

**Comment by DZGoldman** [comment]
@WietseWind  we specifically want % owned by top 100, which currently isn't computable from the `richlist` endpoint.
 
 I'm scraping that exact stat from the front-end now; could you just directly expose it via the api? (/ is it already somewhere?)

**Comment by WietseWind** [comment]
@DZGoldman You can use this endpoint for that;
https://ledger.exposed/api/wallet-toplist/100/0 - The URL is formatted as take/skip, so this URL is: start at `0`, and take `100` accounts. You can sum `/Balance` and compare it with the coin totals from the **Richlist** endpoint.

**Comment by DZGoldman** [comment]
all set.
@ummjackson I used `requests`, not sure if it's installed on the server (maybe we wanna add a `requirements.txt`?)

**Comment by WietseWind** [comment]
@DZGoldman Almost; [the fallback](https://github.com/ummjackson/awdy/blob/9214a7f7fa886f4dd97f7074c50f6bf539c1e4c9/_data/scripts/ripple.py#L49) uses a dynamic attribute (`data-v-ca88cbc2`) assigned when building the Vue app. This attribute will change at new build.

**Comment by WietseWind** [comment]
I'll add an #id at next deploy 👍

File: _data/scripts/ripple.py [hunk 2]
```diff
@@ -16,18 +16,39 @@ def get_public_nodes(self):
...
+            # convert to %
+            wealth_distribution = round(wealth_distribution * 100)
+            print('XRP wealth distribution', wealth_distribution)
+            return wealth_distribution
+            
+        except:   
+            print('XRP api failed (ledger exposed), falling back to web crawler:') 
+            # in case api is down but site is still up? unlikely, but may as well try:
+            self.get_page("https://ledger.exposed/rich-stats");
+            wealth_distribution =  self.attempt_find_element( lambda:self.driver.find_element_by_css_selector('[data-v-ca88cbc2].large > b')).text
```
Comment (WietseWind) [inline _data/scripts/ripple.py]: @DZGoldman You can use the selector `'#top_percentage'` in a few minutes 👍 
image

**Comment by ummjackson** [comment]
Hey all, this is progressing nicely! What I want to have before I merge this into the repo is the data collection automated for _all_ the assets... I know that's a lot but maybe we can crowdsource from the communities? 

Once that's done, I want to test a deployment which updates the YAML files and then pushes a commit daily with the refreshed numbers.

**Comment by DZGoldman** [comment]
Great! Pushed up empty class-templates for all the remaining coins — will see if we can crowd-source some help.

YAML writing should be working; committing still needs to be implemented, although my instinct would be to unversion the yaml files and just update them directly on the server itself (though maybe you have something more in mind here.)

**Comment by ummjackson** [comment]
Excellent, thanks for the updates @DZGoldman! The whole thing is actually hosted on GitHub pages using Jekyll, so when it does it's automatic build process it loops through the YAML files per this template (https://github.com/ummjackson/awdy/blob/master/index.html) and spits out the content.

Because of this, the best approach would be:

1. Run the Python scripts automatically updating the JSON
2. Commit the YAML files to the repo
3. GitHub Pages / Jekyll auto builds and refreshes the site

I can figure out the commit stuff as that'll likely need to involve repo API keys etc.

**Comment by MarkusTeufelberger** [comment]
You could just let the other coins return static numbers for now and add them in the future?

Also rippples are now called "XRP" in your files, so you might want to change the python file name too.

File: _data/scripts/ripple.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+    def get_public_nodes(self):
+        try:
+            # first try API
+            res = requests.get('https://data.ripple.com/v2/network/topology/')
+            public_nodes_count = res.json()['node_count']
+            assert(isinstance(public_nodes_count, int))
+            return public_nodes_count
+        except:
+            print('XRP api failed (xrp charts), falling back to web crawler:')
```
Comment (MarkusTeufelberger) [inline _data/scripts/ripple.py]: Xrpcharts gets its data from the API endpoint that failed at this point in the code - it is likely to return no or bogus data too.

You could implement a crawler yourself that recursively scrapes peer lists of all publicly reachable nodes, but that would be a bit overkill in my opinion.

**Comment by DZGoldman** [comment]
Sweet — 14/24 of the coins are done. I'll pick it up again this week and see if I can finish it off.

**Comment by DZGoldman** [comment]
Update — current status: https://github.com/ummjackson/awdy/issues/44#issuecomment-430844896

**Pull request closed without merging**